File: src/solver/sat/DPLL.java
package solver.sat;

import solver.sat.BranchingStrategies.BranchingStrategy;

import java.util.*;

public class DPLL {

    BranchingStrategy branchingStrategy;

    public DPLL(BranchingStrategy branchingStrategy) {
        this.branchingStrategy = branchingStrategy;
    }

//    private Set<Integer> findPureSymbols(SATInstance instance) {
//        // TODO: write this!!
//        HashSet<Integer> pureSymbols = new HashSet<>();
//        HashSet<Integer> removed = new HashSet<>();
//
//        for (Set<Integer> clause : instance.clauses) {
//            for (Integer var : clause) {
//                if (removed.contains(var) || removed.contains(-var)) {
//                    continue;
//                } else if (pureSymbols.contains(-var)) {
//                    pureSymbols.remove(-var);
//                    removed.add(-var);
//                } else pureSymbols.add(var);
//            }
//        }
//
//        return pureSymbols;
//    }

    private void propagatePureSymbols(Set<Integer> pureSymbols, SATInstance instance, Model model) {
        // System.out.println("pure symbols propagating: " + pureSymbols);
        List<Set<Integer>> updatedClauses = new ArrayList<>();
        for (Set<Integer> clause : instance.clauses) {
            boolean keepClause = true;
            for (Integer pureSymbol : pureSymbols) {
                if (clause.contains(pureSymbol)) {
                    instance.numClauses--;
                    keepClause = false;
                    break;
                }
            }
            if (keepClause) {
                updatedClauses.add(clause);
            } else {
                for (Integer literal : clause) {
                    instance.reduceLiteralCount(literal);
                }
            }
        }
        for (Integer pureSymbol : pureSymbols) {
            // if (pureSymbol == -364) {
            //     System.out.println("-364 literal count: " + instance.literalCounts.get(-364));
            //     System.out.println("364 literal count: " + instance.literalCounts.get(364));
            // }
            // if (model.model.contains(pureSymbol))
            //     System.out.println("ERROR: pure symbol " + pureSymbol + " already in model!");
            // if (model.model.contains(-pureSymbol))
            //     System.out.println("ERROR: pure symbol negation " + (-pureSymbol) + " already in model!");
            // if (instance.literalCounts.containsKey(pureSymbol))
            //     System.out.println("removed pure literal " + pureSymbol + " has non-zero literal count: " + instance.literalCounts.get(pureSymbol));
            // if (instance.literalCounts.containsKey(-pureSymbol))
            //     System.out.println("supposedly non-existent negated pure literal " + (-pureSymbol) + " has non-zero literal count: " + instance.literalCounts.get(-pureSymbol));
        }
        model.model.addAll(pureSymbols); // TODO: create a setModel method
        instance.clauses = updatedClauses; // TODO: maybe create a setClauses method
        for (Integer pureSymbol : pureSymbols) {
            instance.numVars --;
            instance.vars.remove((pureSymbol < 0) ? -1 * pureSymbol : pureSymbol);
        }
    }

//    private Integer findUnitClause(SATInstance instance) {
//        // TODO: maybe find all unit clauses?? and propagate them all??
//        // NOTE: returns 0 if no unit clause is found
//        for (Set<Integer> clause : instance.clauses) {
//            if (clause.size() == 1) {
//                for (Integer var : clause) {
//                    return var;
//                }
//            }
//        }
//        return 0;
//    }
    private void findUnitClauses(SATInstance instance) {
        for (Set<Integer> clause : instance.clauses) {
            if (clause.size() == 1) {
                for (Integer literal : clause) {
                    instance.unitClauses.add(literal);
                    break;
                }
            }
        }
    }

    private void propagateUnitClause(SATInstance instance, Integer literal, Model model) throws EmptyClauseFoundException {
        // System.out.println("unit clause propagating: " + literal);
        List<Set<Integer>> updatedClauses = new ArrayList<>();
        for (Set<Integer> clause : instance.clauses) {
            if (clause.contains(literal)) {
                // if (literal == -154) {
                //     System.out.println("clause: " + clause);
                // }
                for (Integer clauseLiteral : clause) {
                    instance.reduceLiteralCount(clauseLiteral);
                    if (clauseLiteral.equals(literal) || clauseLiteral.equals(-literal))
                        continue;
                    // if (literal == -154) {
                    //     System.out.println("clause: " + clause);
                    //     System.out.println("clause literal: " + clauseLiteral);
                    //     System.out.println("literal counts of clause literal: " + instance.literalCounts.get(clauseLiteral));
                    //     System.out.println("literal counts of -clause literal: " + instance.literalCounts.get(-clauseLiteral));
                    // }
                    if (!instance.literalCounts.containsKey(clauseLiteral) && instance.literalCounts.containsKey(-clauseLiteral)) {
                        // if (literal == -154) {
                        //     System.out.println("clause: " + clause);
                        //     System.out.println("clause literal: " + clauseLiteral);
                        //     System.out.println("literal counts of clause literal: " + instance.literalCounts.get(clauseLiteral));
                        //     System.out.println("literal counts of -clause literal: " + instance.literalCounts.get(-clauseLiteral));
                        // }
                        if (instance.unitClauses.contains(-clauseLiteral) || instance.unitClauses.contains(clauseLiteral))
                            continue;
                        instance.pureSymbols.add(-clauseLiteral);
                    }
                }
            }
            else if (clause.contains(-literal)) {
                // if (literal == -364) {
                //     System.out.println("clause: " + clause);
                //     System.out.println("unit literal being propagated: " + literal);
                // }
                clause.remove(-literal);
                instance.reduceLiteralCount(-literal);
                if (clause.isEmpty()) {
                    throw new EmptyClauseFoundException("empty clause found!");
                }
                updatedClauses.add(clause);
                if (clause.size() == 1) {
                    for (Integer unitLiteral : clause) {
                        if (!instance.pureSymbols.contains(unitLiteral))
                            instance.unitClauses.add(unitLiteral);
                        break;
                    }
                }
            } else {
                updatedClauses.add(clause);
            }
        }
        // if (model.model.contains(literal))
        //     System.out.println("ERROR: unit clause " + literal + " already in model!");
        // if (model.model.contains(-literal))
        //     System.out.println("ERROR: unit clause negation " + (-literal) + " already in model!");
        model.model.add(literal); // TODO: create a setModel method
        instance.clauses = updatedClauses; // TODO: maybe create a setClauses method
        instance.numClauses = updatedClauses.size();
        instance.vars.remove(literal < 0 ? -literal : literal);
        instance.numVars--;
    }

//    private Integer pickBranchVariable(SATInstance instance) throws NoVariableFoundException {
//        // TODO: make smarter search heuristics at some point
//        for (Integer var : instance.vars) {
//            return var;
//        }
//        throw new NoVariableFoundException("No variable found in SATInstance: " + instance);
//    }

    private boolean hasEmptyClause(SATInstance instance) {
        for (Set<Integer> clause : instance.clauses) {
            if (clause.isEmpty()) {
                return true;
            }
        }
        return false;
    }

    private boolean isSAT(SATInstance instance) {
        // checks if guaranteed to be sat already
        return instance.clauses.size() == 0;
    }

    private  DPLLResult dpllInternal(SATInstance instance, Model model) {
//        System.out.println("num clauses: " + instance.clauses.size());
//        System.out.println("num unit clauses: " + instance.unitClauses.size());
//        System.out.println("num pure literals: " + instance.pureSymbols.size());
        if (isSAT(instance)) {
            return new DPLLResult(instance, model, true);
        }

//        if (hasEmptyClause(instance)) {
//            return new DPLLResult(instance, model, false);
//        }

//        // TODO: mess with order of pure/unit symbols
//        Set<Integer> pureSymbols = findPureSymbols(instance);
//        if (!pureSymbols.isEmpty()) {
//            propagatePureSymbols(pureSymbols, instance, model);
//            return dpll(instance, model);
//        }
        if (!instance.pureSymbols.isEmpty()) {
            propagatePureSymbols(instance.pureSymbols, instance, model);
            instance.pureSymbols.clear();
            return dpllInternal(instance, model);
        }

        try {
            if (!instance.unitClauses.isEmpty()) {
                // if (instance.unitClauses.contains(-364)) {
                //     System.out.println("all unit clauses: " + instance.unitClauses);
                // }
                Integer unitLiteral = 0;
                for (Integer literal : instance.unitClauses) {
                    unitLiteral = literal;
                    break;
                }
                instance.unitClauses.remove(unitLiteral);
                propagateUnitClause(instance, unitLiteral, model);
                return dpllInternal(instance, model);
            }
        } catch (EmptyClauseFoundException e) { // UNSAT
            return new DPLLResult(instance, model, false);
        }

        try {
            Integer branchVariable = this.branchingStrategy.pickBranchingVariable(instance);
            // if (branchVariable == 364)
            //     System.out.println("branching on 364");
            // if (branchVariable == -364)
            //     System.out.println("branching on -364");
            // if (model.model.contains(branchVariable))
            //     System.out.println("branch Variable " + branchVariable + " already in model");
            // if (model.model.contains(-branchVariable))
            //     System.out.println("negated branch variable " + (-branchVariable) + " already in model");

            // positive assumption
            SATInstance positiveInstance = instance.copy();
            Model positiveModel = model.copy();
            Set<Integer> positiveUnitClause = new HashSet<>();
            positiveUnitClause.add(branchVariable);
            positiveInstance.addClause(positiveUnitClause);
            positiveInstance.literalCounts.put(branchVariable, positiveInstance.literalCounts.getOrDefault(branchVariable, 0) + 1);
            positiveInstance.numClauses ++;
            positiveInstance.unitClauses.add(branchVariable);
//            propagateUnitClause(positiveInstance, branchVariable, positiveModel);

            DPLLResult positiveAssumptionResult = dpllInternal(positiveInstance, positiveModel);
            if (positiveAssumptionResult.isSAT) {
                return positiveAssumptionResult;
            }

            // negative assumption
            SATInstance negativeInstance = instance.copy();
            Model negativeModel = model.copy();
            Set<Integer> negativeUnitClause = new HashSet<>();
            negativeUnitClause.add(-1 * branchVariable);
            negativeInstance.addClause(negativeUnitClause);
            negativeInstance.literalCounts.put(-branchVariable, negativeInstance.literalCounts.getOrDefault(-branchVariable, 0) + 1);
            negativeInstance.numClauses ++;
            negativeInstance.unitClauses.add(-branchVariable);
//            propagateUnitClause(negativeInstance, -1 * branchVariable, negativeModel);

            return dpllInternal(negativeInstance, negativeModel);
        }
        catch (NoVariableFoundException e) {
            System.out.println(e.getMessage());
            return null;
        }

    }

    public DPLLResult dpll(SATInstance instance, Model model) {
        if (hasEmptyClause(instance)) {
            return new DPLLResult(instance, model, false);
        }

        // finding pure symbols
        for (Set<Integer> clause : instance.clauses) {
            for (Integer literal : clause) {
                Integer literalCount = instance.literalCounts.getOrDefault(literal, 0);
                instance.literalCounts.put(literal, literalCount + 1);
                if (instance.literalCounts.containsKey(-literal)) {
                    instance.pureSymbols.remove(literal);
                    instance.pureSymbols.remove(-literal);
                } else {
                    instance.pureSymbols.add(literal);
                }
            }
        }

        propagatePureSymbols(instance.pureSymbols, instance, model);
        instance.pureSymbols.clear();

        findUnitClauses(instance);

        return this.dpllInternal(instance, model);
    }
}

File: src/solver/sat/DPLL.cpp
#include <iostream>
#include <cassert>
#include "include/DPLL.hpp"
#include "include/utils.hpp"

/**
 * Class that runs the DPLL algorithm on a given SATInstance
*/
DPLL::DPLL(BranchingStrategy* branchingStrategy, SATInstance* instance, Model* model) {
    this->branchingStrategy = branchingStrategy;
    this->instance = instance;
    this->model = model;

    this->remainingClauses = new std::unordered_set<int>();
    this->removedLiterals = new std::unordered_map<int, std::unordered_set<int>*>();
    this->removedClauseStack = new std::vector<std::unordered_set<int>*>();
    this->removedLiteralStack = new std::vector<std::unordered_map<int, std::unordered_set<int>*>*>();
    this->assignmentStack = new std::vector<std::unordered_set<int>*>();

    branchingStrategy->setContext(this->remainingClauses, this->removedLiterals);
}

/**
 * Returns the variable for this literal
*/
int getVar(int literal) {
    return literal < 0 ? -literal : literal;
}

/**
 * Propagates all pure symbols found so far
*/
void DPLL::propagatePureSymbols() {
    std::unordered_set<int> *filteredPureSymbols = new std::unordered_set<int>();
    // exclude any pure symbols whose count is 0 (should be none like this -- extra
    // safety check)
    for (int pureSymbol : *(this->instance->pureSymbols)) {
        if (instance->getLiteralCounts(pureSymbol)->at(getVar(pureSymbol) - 1) > 0)
            filteredPureSymbols->insert(pureSymbol);
    }
    delete(instance->pureSymbols); // delete old memory on the heap
    instance->pureSymbols = filteredPureSymbols;

    std::unordered_set<int> *newPureSymbols = new std::unordered_set<int>(); // newly created pure symbols go here
    std::vector<int> clausesToRemove; // clauses that need to be removed
    for (int clauseIdx : *this->remainingClauses) { // only loop through remaining clauses
        std::unordered_set<int> *clause = this->instance->clauses->at(clauseIdx);
        bool keepClause = true;
        for (int pureSymbol : *instance->pureSymbols) {
            if (setContains(clause, pureSymbol)) { // if clause contains a pure symbol, remove the clause
                keepClause = false;
                break;
            }
        }
        if (!keepClause) { // this clause needs to be removed
            std::unordered_set<int> emptySet;
            std::unordered_set<int> *clauseRemovedLiterals = getOrDefault(this->removedLiterals, clauseIdx, &emptySet);
            for (int literal : *clause) {
                // reduce the counts of all literals in the clause (other than literals that are already removed)
                if (setContains(clauseRemovedLiterals, literal))
                    continue;
                this->instance->reduceLiteralCount(literal);
                int literalVar = getVar(literal);
                // check if a new pure symbol has been created as a result of removing this literal
                if (instance->getLiteralCounts(-literal)->at(literalVar - 1) > 0 && instance->getLiteralCounts(literal)->at(literalVar - 1) == 0) {
                    newPureSymbols->insert(-literal);
                }
            }
            clausesToRemove.push_back(clauseIdx);
            this->removedClauseStack->at(this->removedClauseStack->size() - 1)->insert(clauseIdx);
        }
    }

    // update assignments in model
    for (int pureSymbol : *this->instance->pureSymbols) {
        this->model->model->insert(pureSymbol);
        this->assignmentStack->at(this->assignmentStack->size() - 1)->insert(pureSymbol);
    }
    
    delete(this->instance->pureSymbols); // delete filteredPureSymbols
    this->instance->pureSymbols = newPureSymbols;
    for (int clauseIdx : clausesToRemove) {
        this->remainingClauses->erase(clauseIdx); // update remaining clauses
    }
}

/**
 * Propagates the given literal as a unit clause
*/
void DPLL::propagateUnitClause(int literal) {
    std::unordered_set<int> empty_set;
    std::unordered_set<int> clausesToRemove; // used to update remaining clauses after unit propagation
    bool emptyClauseFound = false;
    
    for (int clauseIdx : *(this->remainingClauses)) {
        std::unordered_set<int> *clause = this->instance->clauses->at(clauseIdx);
        // remove clauses that contain the literal being propagated
        if (setContains(clause, literal)) {
            // get literals removed from this clause
            std::unordered_set<int> *clauseRemovedLiterals = getOrDefault(this->removedLiterals, clauseIdx, &empty_set);
            for (int clauseLiteral : *clause) {
                // if the clauseLiteral has been removed from the clause, do nothing
                if (setContains(clauseRemovedLiterals, clauseLiteral)) {
                    continue;
                }
                
                // decrease literal count for clauseLiteral because its clause is being removed
                this->instance->reduceLiteralCount(clauseLiteral);
                // go to next iteration if clauseLiteral or its negation are already tracked as unit clauses
                if (setContains(this->instance->unitClauses, -clauseLiteral) || setContains(this->instance->unitClauses, clauseLiteral)) {
                    continue;
                }
                
                int clauseLiteralVar = getVar(clauseLiteral);
                // check if new pure symbol is created
                if (this->instance->getLiteralCounts(clauseLiteral)->at(clauseLiteralVar - 1) == 0 && this->instance->getLiteralCounts(-clauseLiteral)->at(clauseLiteralVar - 1) > 0
                    && (literal != clauseLiteral) && (literal != -clauseLiteral)) {
                    this->instance->pureSymbols->insert(-clauseLiteral);
                }
            }
            
            // record (locally and globally) clauses removed by unit propagation
            clausesToRemove.insert(clauseIdx);
            this->removedClauseStack->at(this->removedClauseStack->size() - 1)->insert(clauseIdx);
        } else if (setContains(clause, -literal)) {
            this->instance->reduceLiteralCount(-literal);
            // update removed literals
            if (!mapContainsKey(this->removedLiterals, clauseIdx)) {
                (*this->removedLiterals)[clauseIdx] = new std::unordered_set<int>();
            }
            this->removedLiterals->at(clauseIdx)->insert(-literal);
            // update removedLiterals in its stack
            if (!mapContainsKey(this->removedLiteralStack->at(removedLiteralStack->size() -  1), clauseIdx)) {
                (*this->removedLiteralStack->at(removedLiteralStack->size() - 1))[clauseIdx] = new std::unordered_set<int>();
            }
            this->removedLiteralStack->at(removedLiteralStack->size() - 1)->at(clauseIdx)->insert(-literal);

            // check if empty clause
            if (clause->size() == this->removedLiterals->at(clauseIdx)->size()) {
                emptyClauseFound = true;
                break;
            }

            // if new unit clause created
            if (clause->size() - this->removedLiterals->at(clauseIdx)->size() == 1) {
                for (int unitLiteral : *clause) {
                    if (setContains(removedLiterals->at(clauseIdx), unitLiteral)) 
                        continue;
                    if (!setContains(this->instance->pureSymbols, unitLiteral)) 
                        this->instance->unitClauses->insert(unitLiteral);
                    break;
                }
            }
        }
    }

    // update remaining clauses
    for (int clauseIdx : clausesToRemove) {
        this->remainingClauses->erase(clauseIdx);
    }
    // update assignments
    this->model->model->insert(literal);
    this->assignmentStack->at(assignmentStack->size() - 1)->insert(literal);
    if (emptyClauseFound) {
        throw std::runtime_error("Empty clause found!");
    }

}

/**
 * Checks if a SAT configuration has been reached
*/
bool DPLL::isSAT() {
    // sat if there are no clauses left to be satisfied
    return this->remainingClauses->size() == 0;
}

/**
 * Checks if the instance (initially) contains an empty clause (trivially UNSAT)
*/
bool DPLL::initiallyHasEmptyClause() {
    for (std::unordered_set<int>* clause : *(this->instance->clauses)) {
        if (clause->empty()) {
            return true;
        }
    }
    return false;
}

/**
 * Finds the unit clauses initially present in the instance
*/
void DPLL::findInitialUnitClauses() {
    std::unordered_set<int> emptySet;
    for (int clauseIdx : *(this->remainingClauses)) {
        std::unordered_set<int>* clause = this->instance->clauses->at(clauseIdx);
        std::unordered_set<int>* clauseRemovedLiterals = getOrDefault(this->removedLiterals, clauseIdx, &emptySet);
        
        if (clause->size() - clauseRemovedLiterals->size() == 1) { // if the number of literals left in the clause is 1, it is a unit clause
            for (int literal : *clause) {
                if (setContains(clauseRemovedLiterals, literal)) // skip if this literal has been removed
                    continue;
                if (!setContains(this->instance->pureSymbols, literal)) // if already in pure symbols, don't add again (will be propagated anyways)
                    this->instance->unitClauses->insert(literal);
                break;
            }
        }
    }
}

/**
 * Performs backtracking
*/
void DPLL::backtrack() {
    // pops off the last elements from the stacks
    std::unordered_set<int> *stackRemovedClauses = this->removedClauseStack->back();
    this->removedClauseStack->pop_back();
    std::unordered_map<int, std::unordered_set<int>*> *stackRemovedLiterals = this->removedLiteralStack->back();
    this->removedLiteralStack->pop_back();
    std::unordered_set<int> *stackAssignments = this->assignmentStack->back();
    this->assignmentStack->pop_back();

    // undo the removed clauses (undo clauses removed from instnce since last branch)
    for (int clauseIdx : *stackRemovedClauses) { // loop through all clauses removed since the branch being backtracked
        if (setContains(this->remainingClauses, clauseIdx)) {
            std::cout << "UNEXPECTED: remainingClauses contains clauses that are removed according to the stack!" << std::endl;
        }
        this->remainingClauses->insert(clauseIdx); // the clause is no longer removed

        std::unordered_set<int> emptySet;
        std::unordered_set<int> *clauseRemovedLiterals = getOrDefault(this->removedLiterals, clauseIdx, &emptySet);
        for (int literal : *(this->instance->clauses->at(clauseIdx))) {
            // ignore literals if they are in removedLiterals.
            // if something is in removedLiterals, these are literals that were removed from the clause before
            // the clause itself was removed (when removing a clause as a whole, we don't update removedLiterals for the clause).
            // we want to make sure not to unroll the removal of literals from the clause that took place
            // before the branch being undone.
            if (setContains(clauseRemovedLiterals, literal))
                continue;
            this->instance->increaseLiteralCount(literal);
        }
    }

    // undo the removed literals (undo literals removed from clauses since the last branch)
    for (const auto& pair : *stackRemovedLiterals) {
        int clauseIdx = pair.first;
        std::unordered_set<int> *clauseRemovedLiterals = this->removedLiterals->at(clauseIdx);
        for (int removedLiteral : *pair.second) {
            if (!setContains(clauseRemovedLiterals, removedLiteral))
                std::cout << "UNEXPECTED: removedLiterals doesn't contain literal that was removed according to the stack!" << std::endl;
            clauseRemovedLiterals->erase(removedLiteral); // this literal is no longer removed from this clause
            this->instance->increaseLiteralCount(removedLiteral); // update the count for the literal 
        }
    }

    // undo the assignments made since thhe last branch
    for (int assignment : *stackAssignments) {
        if (!setContains(this->model->model, assignment))
            std::cout << "UNEXPECTED: model doesn't contain all assignments stored in the stack!" << std::endl;
    }
    for (int assignment : *stackAssignments) {
        this->model->model->erase(assignment);
    }
    
    // now that we have backtracked, no unit clauses or pure symbols found here are valid anymore
    this->instance->unitClauses->clear();
    this->instance->pureSymbols->clear();

    // clean up memory created on the heap:
    delete(stackRemovedClauses);
    for (const auto& pair : *stackRemovedLiterals) {
        delete(pair.second);
    }
    delete(stackRemovedLiterals);
    delete(stackAssignments);
}


/**
 * Iteratively runs DPLL
*/
DPLLResult* DPLL::dpllInternal() {
    // initialize stack of branching literals
    std::vector<int> dpllLiteralStack;
    // indicates whether the negation of the literal at same index in
    // dpllLiteralStack has been used in branching
    std::vector<bool> dpllNegCheckedStack;
    // stores whether the current branch has evaluated to unsat
    bool unsatReached = false;

    while (!isSAT()) {
        if (unsatReached) {
            // return UNSAT if all branches checked
            if (dpllLiteralStack.empty())
                return new DPLLResult(this->instance, this->model, false);
            
            // if negative branch checked, remove literal from stack and backtrack again
            bool negativeBranchChecked = dpllNegCheckedStack.back();
            if (negativeBranchChecked) {
                dpllLiteralStack.pop_back();
                dpllNegCheckedStack.pop_back();
                continue;
            }
            // if negative branch not yet checked, backtrack and run DPLL on
            // negation of branching variable 
            backtrack();
            dpllLiteralStack.back() *= -1;
            instance->unitClauses->insert(dpllLiteralStack.back());
            dpllNegCheckedStack.back() = true;
            unsatReached = false;
            continue;
        }

        // propagate pure symbols
        if (!this->instance->pureSymbols->empty()) {
            propagatePureSymbols();
            continue;
        }

        try {
            // propagate a single unit clause
            if (!this->instance->unitClauses->empty()) {
                int unitLiteral = 0;
                for (int unitClause : *this->instance->unitClauses) {
                    unitLiteral = unitClause;
                    break;
                }

                this->instance->unitClauses->erase(unitLiteral);
                propagateUnitClause(unitLiteral);
                continue;
            }
        } catch (...) {
            unsatReached = true;
            continue;
        }

        // we need to branch!
        try {
            int branchVariable = this->branchingStrategy->pickBranchingVariable(this->instance);
            dpllLiteralStack.push_back(branchVariable);
            dpllNegCheckedStack.push_back(false);

            // make new entries on the global stacks
            this->removedClauseStack->push_back(new std::unordered_set<int>());
            this->removedLiteralStack->push_back(new std::unordered_map<int, std::unordered_set<int>*>());
            this->assignmentStack->push_back(new std::unordered_set<int>());

            // mark the branching variable as a unit clause that needs to be propagated
            this->instance->unitClauses->insert(branchVariable);
            continue;
        } catch (const std::exception& e) {
            std::cerr << "Exception caught: " << e.what() << std::endl;
            return NULL;
        }

    }
    
    // we have found a SAT assignment (exited the loop)
    return new DPLLResult(this->instance, this->model, true);
}

/**
 * Sets up state for dpll algorithm (which will be run in dpllInternal, above)
*/
DPLLResult* DPLL::dpll() {
    if (initiallyHasEmptyClause()) { // if trivially UNSAT  (instance initially has empty clause), return UNSAT
        return new DPLLResult(this->instance, this->model, false);
    }
    
    // initialize stacks with empty elements (these initial elements should always remain on the stack -- never used
    // while backtracking etc since these are from before we ever branch)
    this->removedClauseStack->push_back(new std::unordered_set<int>());
    this->removedLiteralStack->push_back(new std::unordered_map<int, std::unordered_set<int>*>());
    this->assignmentStack->push_back(new std::unordered_set<int>());
    
    // populate remainingClauses -- put all clauses there (all clauses are currently unsatisfied)
    for (int i = 0; i < this->instance->clauses->size(); i++) {
        this->remainingClauses->insert(i);
    }
    assert(this->remainingClauses->size() == this->instance->clauses->size()); // safety check -- all clauses should be remaining

    // find initially occurring pure symbols and populate literal counts
    for (std::unordered_set<int> *clause : *this->instance->clauses) {
        for (int literal : *clause) {
            this->instance->increaseLiteralCount(literal);
            if (this->instance->getLiteralCounts(-literal)->at(getVar(-literal) - 1) > 0) {
                // if literal and -literal both exist, make sure neither is marked as a pure symbol
                this->instance->pureSymbols->erase(literal);
                this->instance->pureSymbols->erase(-literal);
            } else { // if -literal doesn't exist, literal is a pure symbol
                this->instance->pureSymbols->insert(literal);
            }
        }
    }

    propagatePureSymbols(); // propagate the pure symbols

    findInitialUnitClauses(); // find the initial unit clauses (will be propagated within dpllInternal)
    
    return dpllInternal(); // runs DPLL
}
